### DISCUSSION TITLE
Custom debug adapter within the browser context

### DISCUSSION THREAD

**Discussion by pchuong** [eclipse-theia/theia#12332] [Q&A]
Hello,
I am trying to find out if there is a way to create a debug adapter contribution that communciate with the debug server that runs in the browser context. From looking at the code, the debug-adapter-session-manager.ts file supports a remote websocket adapter and an executable adapter. Is there any option that can be use to provide a custom adapter that conforms to the DAP, and let the custom adapter manage the communicate to the backend debug server?

Thank you!

**Comment by msujew** [comment]
@thegecko I believe you have a bit of knowledge on that, since you build a frontend based debug adapter.

**Comment by thegecko** [comment]
@pchuong you can create a debug adapter which runs in the browser. Depending on what you are debugging, this may or may not be what you want, though!

Will your debugger run on the backend? Or are you trying to set up an adapter which debugs something running in a remote container?

**Comment by pchuong** [comment]
Hi @thegecko, my debugger is running on the backend, it runs on the user's local PC. It uses a websocket for communication . There is existing proprietary debugger API that is available in the browser context, the same API is also used by other applications. I like to be able to adapt Theia DAP to the API already existed in the browser.

Is there any example code or documentation that you can point me to?

Thanks,
Patrick

**Comment by thegecko** [comment]
I recommend you look at the mock debug example.

https://github.com/microsoft/vscode-mock-debug

This can be run in the frontend, but what you'll need to do is write an inlineadapter which streams the commands to and from a websocket.

something along the lines of:

```typescript

import * as vscode from 'vscode';
import * as websocket from 'websocket-stream';
import { DebugProtocol } from 'vscode-debugprotocol';
import { ProtocolServer } from 'vscode-debugadapter/lib/protocol';

class InlineDebugSession extends ProtocolServer {
    protected adapter: WebsocketDebugAdapter;

    public constructor(address: string, protected pathMapping?: { [key: string]: string }) {
        super();
        this.adapter = new WebsocketDebugAdapter(address);
        this.adapter.on('message', message => this.onMessage(message));
    }

    protected async dispatchRequest(request: DebugProtocol.Request): Promise<void> {
        this.adapter.send(request);
    }

    protected onMessage(message: DebugProtocol.ProtocolMessage): void {
        if (message.type === 'response') {
            const msg = message as DebugProtocol.Response;
            // sendResponse expects seq not to be set as it will add it's own
            msg.seq = 0;
            this.sendResponse(msg);
        } else if (message.type === 'event') {
            this.sendEvent(message as DebugProtocol.Event);
        }
    }
}

class WebsocketDebugAdapter extends ProtocolServer {
    protected stream: NodeJS.ReadWriteStream;

    public constructor(address: string) {
        super();
        this.stream = this.createStream(address);
        super.start(this.stream, this.stream);
    }

    public handleMessage(message: DebugProtocol.ProtocolMessage): void {
        this.emit('message', message);
    }

    public send(message: DebugProtocol.ProtocolMessage) {
        const json = JSON.stringify(message);
        this.stream.write(`Content-Length: ${Buffer.byteLength(json, 'utf8')}\r\n\r\n${json}`, 'utf8');
    }

    protected createStream(address: string): NodeJS.ReadWriteStream {
        return websocket(address, {
            handshakeTimeout: 10000
        });
    }
}

export class DebugFactory implements vscode.DebugAdapterDescriptorFactory {
    public async createDebugAdapterDescriptor(session: vscode.DebugSession, _executable: vscode.DebugAdapterExecutable | undefined): Promise<vscode.DebugAdapterDescriptor | undefined> {
        const ssession = new InlineDebugSession(session.configuration.address);
        return new vscode.DebugAdapterInlineImplementation(ssession);
    }
}

```

**Reply by pchuong** [reply]
I have browsed the Theia code to see how Debug Adapter is instantiated by the plugin-ext extension, but I don't fully understand how everything is connected together. From my limited knowledge, it looks like the Debug Adapter is instantiated from the backend and debug messages are proxied to the frontend vscode extension. 

My use cases are to have an IDE to support both desktop and cloud. There is no issue when both the frontend and backend code are running in the same PC. The question is, will performance be an issue when the backend is running on a remote server that is hosted somewhere else. For example, user opens a browser in Asia and connect to a server running in North America. Will all the debug command traffic has to be send between the user's PC and the remote server, i.e step, run, read memory, etc...

Ideally, I would like to have a way to intercept the Debug commands that are send from the frontend and redirect it locally to a webserver that is running on the user's local machine. Is this something that is do-able?

**Reply by msujew** [reply]
> Ideally, I would like to have a way to intercept the Debug commands that are send from the frontend and redirect it locally to a webserver that is running on the user's local machine. Is this something that is do-able?

@pchuong That sounds quite over-engineered, and I'm not even sure whether that would improve the user experience. Since most of Theia's work is done on the frontend side, even a ping of a few hundred ms (not unusual for Asia <-> NA) shouldn't impact the user experience by a lot.

If user ping becomes a problem, I believe the "easier" fix would be to use a closer data center.

**Reply by pchuong** [reply]
Hi @msujew,
Using Asia and NA might be an extreme case. I am building a debugger for embedded devices. When reading memory, disassemble target instructions, sending trace data. Data that is transfer from client browser to a remote server and then back to the client machine can be problematic. Since the target device is sitting beside the users host PC, it would make sense if there is a way to communicate directly from the Theia browser to a local process,  while the front is served by the remote web server.

**Reply by msujew** [reply]
I see, I might have misunderstood your original idea. Right, so in my mind, you would have something like a "helper app" that talks with the browser via some protocol - websocket seems to be pretty useful for that. In the end all that this requires is a debug adapter running in the browser plugin host (as outlined by Rob) communicating with the local app which is in charge of communicating with the embedded device.

That seems pretty feasible.

**Comment by pchuong** [comment]
Thank you @thegecko, that is very helpful. I'll give it a try.
